File: lib/src/KeyServerClient.php
<?php
declare(strict_types=1);

namespace VaultCode;

/**
 * Client HTTP del key-server: firma HMAC per-installazione + nonce + timestamp
 * (anti-replay), su TLS. Verifica la lista revoche firmata Ed25519.
 */
final class KeyServerClient implements KeyServerInterface
{
    private Config $cfg;

    public function __construct(Config $cfg)
    {
        $this->cfg = $cfg;
    }

    private function signedHeaders(string $method, string $path, string $body): array
    {
        $nonce = bin2hex(random_bytes(16));
        $ts = (string)time();
        $msg = Crypto::buildCanonicalMessage($method, $path, $this->cfg->installUuid, $nonce, $ts, $body);
        $sig = Crypto::computeSignature($this->cfg->installSecret, $msg);
        return [
            'Content-Type: application/json',
            'X-Install-UUID: ' . $this->cfg->installUuid,
            'X-Nonce: ' . $nonce,
            'X-Timestamp: ' . $ts,
            'X-Signature: ' . $sig,
        ];
    }

    /** @return array{code:int,body:string} */
    private function http(string $method, string $path, array $headers, string $body): array
    {
        $ch = curl_init($this->cfg->keyServerUrl . $path);
        curl_setopt_array($ch, [
            CURLOPT_CUSTOMREQUEST => $method,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_HTTPHEADER => $headers,
            CURLOPT_TIMEOUT => (int)ceil($this->cfg->timeout),
            CURLOPT_POSTFIELDS => $method === 'POST' ? $body : null,
        ]);
        // Pinning TLS della chiave pubblica del key-server (anti-MITM/redirect).
        // Un mismatch fa fallire curl → cascata non bloccante (INVARIANTE 7).
        if ($this->cfg->keyServerPin !== '') {
            curl_setopt($ch, CURLOPT_PINNEDPUBLICKEY, $this->cfg->keyServerPin);
        }
        $resp = curl_exec($ch);
        if ($resp === false) {
            $err = curl_error($ch);
            curl_close($ch);
            throw new \RuntimeException("trasporto verso il key-server fallito: $err");
        }
        $code = (int)curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);
        return ['code' => $code, 'body' => (string)$resp];
    }

    public function fetchKey(string $module, int $ck, string $frag): array
    {
        // Opzione B: si richiede la chiave della SOLA porzione (frag_id).
        $body = json_encode([
            'module_id' => $module,
            'ck_version' => $ck,
            'frag_id' => $frag,
            'license_token' => $this->cfg->licenseToken,
        ]);
        $headers = $this->signedHeaders('POST', '/key', $body);
        $r = $this->http('POST', '/key', $headers, $body);
        $data = json_decode($r['body'], true);
        if (!is_array($data)) {
            throw new \RuntimeException("risposta /key non valida (HTTP {$r['code']})");
        }
        $data['_http'] = $r['code'];
        return $data;
    }

    public function postTamper(string $tipo, ?string $dettaglio): void
    {
        $body = json_encode(['tipo' => $tipo, 'dettaglio' => $dettaglio]);
        try {
            $this->http('POST', '/tamper', $this->signedHeaders('POST', '/tamper', $body), $body);
        } catch (\Throwable $e) {
            // Reporting non bloccante: si ignora l'errore (INVARIANTE 7).
        }
    }

    /**
     * Contesto d'installazione (anti-copia §4): DOVE gira l'opera. Dati minimi e
     * dichiarati — dominio servito, IP del server, macchina, percorso. Mai contenuto
     * o attività del cliente. Best-effort: assente o parziale non è mai un errore.
     */
    private function installContext(): array
    {
        $s = $_SERVER ?? [];
        $dominio = (string)($s['HTTP_HOST'] ?? $s['SERVER_NAME'] ?? '');
        $dominio = preg_replace('/:\d+$/', '', $dominio);                  // via la porta
        $hostip  = (string)($s['SERVER_ADDR'] ?? '');                      // IP del server (spesso IPv4)
        $path    = (string)($s['DOCUMENT_ROOT'] ?? __DIR__);
        // machine_id stabile ma NON in chiaro: hash di /etc/machine-id (fallback hostname).
        $mid = @file_get_contents('/etc/machine-id');
        if ($mid === false || $mid === '') { $mid = @php_uname('n'); }
        $machine = $mid ? substr(hash('sha256', (string)$mid), 0, 24) : '';
        return array_filter([
            'dominio'      => $dominio,
            'host'         => $hostip,
            'machine_id'   => $machine,
            'install_path' => $path,
        ], static fn($v) => $v !== '' && $v !== null);
    }

    public function fetchStatus(): ?array
    {
        // Riporta la versione della libreria + il contesto d'installazione (dominio/IP
        // server/macchina) → lo studio mostra "aggiornata?" e DOVE gira l'opera (§4).
        $body = json_encode(['client_version' => Runtime::VERSION] + $this->installContext());
        try {
            $r = $this->http('POST', '/status', $this->signedHeaders('POST', '/status', $body), $body);
        } catch (\Throwable $e) {
            return null; // disponibilità: il chiamante terrà l'ultimo stato noto
        }
        $data = json_decode($r['body'], true);
        if (!is_array($data) || !isset($data['signed_status'])) { return null; }
        $token = (string)$data['signed_status'];
        $payload = Crypto::verifyToken($this->cfg->ed25519PublicKeyPem, $token);
        if ($payload === null) { return null; }                       // firma non valida → ignora
        if (($payload['type'] ?? '') !== 'vaultcode/status/v1') { return null; }
        if (($payload['install_uuid'] ?? '') !== $this->cfg->installUuid) { return null; }
        $payload['_token'] = $token;                                   // per la cache firmata lato Runtime
        return $payload;
    }

    public function fetchRevocations(): ?array
    {
        try {
            $r = $this->http('GET', '/revocations', ['Accept: application/json'], '');
        } catch (\Throwable $e) {
            return null;
        }
        $data = json_decode($r['body'], true);
        if (!is_array($data) || !isset($data['signed_document'])) { return null; }
        $payload = Crypto::verifyToken($this->cfg->ed25519PublicKeyPem, (string)$data['signed_document']);
        if ($payload === null || !isset($payload['revoked'])) { return null; }
        $ids = [];
        foreach ((array)$payload['revoked'] as $rev) {
            if (isset($rev['license_id'])) { $ids[] = (string)$rev['license_id']; }
        }
        return $ids;
    }
}

File: client-lib-python/vaultcode/key_server_client.py
"""Client HTTP del key-server: firma HMAC per-installazione + nonce + timestamp
(anti-replay) su TLS. Verifica le risposte firmate Ed25519. Parità con
``KeyServerClient.php``. Solo stdlib (`urllib`) + `cryptography`.

Nota: il pinning TLS della chiave pubblica (CURLOPT_PINNEDPUBLICKEY in PHP) non è
replicato qui (urllib non lo espone in modo pulito): la difesa primaria contro un
server contraffatto resta la **firma Ed25519** di /status, /revocations, /release.
"""

from __future__ import annotations

import hashlib
import json
import os
import socket
import time
import urllib.request

from . import crypto
from .config import Config

RUNTIME_VERSION = "1.0.1"


def _install_context() -> dict:
    """Contesto d'installazione del bot (anti-copia §4): DOVE gira l'opera. Dati
    minimi e dichiarati (host/IP/macchina/percorso), mai contenuto o attività del
    cliente. Best-effort: ogni campo è opzionale e non solleva mai."""
    ctx: dict[str, str] = {}
    try:
        ctx["dominio"] = socket.getfqdn() or socket.gethostname()       # contesto (no HTTP host nei bot)
    except Exception:  # noqa: BLE001
        pass
    try:
        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        try:
            s.connect(("8.8.8.8", 80)); ctx["host"] = s.getsockname()[0]  # IP locale/uscita (best-effort)
        finally:
            s.close()
    except Exception:  # noqa: BLE001
        pass
    try:
        mid = ""
        for p in ("/etc/machine-id", "/var/lib/dbus/machine-id"):
            if os.path.exists(p):
                mid = open(p, encoding="utf-8").read().strip()
                break
        ctx["machine_id"] = hashlib.sha256((mid or socket.gethostname()).encode()).hexdigest()[:24]
    except Exception:  # noqa: BLE001
        pass
    try:
        ctx["install_path"] = os.getcwd()
    except Exception:  # noqa: BLE001
        pass
    return {k: v for k, v in ctx.items() if v}


class KeyServerClient:
    def __init__(self, cfg: Config):
        self.cfg = cfg

    def _signed_headers(self, method: str, path: str, body: bytes) -> dict:
        nonce = os.urandom(16).hex()
        ts = str(int(time.time()))
        msg = crypto.build_canonical_message(method, path, self.cfg.install_uuid, nonce, ts, body)
        sig = crypto.compute_signature(self.cfg.install_secret, msg)
        return {
            "Content-Type": "application/json",
            "X-Install-UUID": self.cfg.install_uuid,
            "X-Nonce": nonce,
            "X-Timestamp": ts,
            "X-Signature": sig,
        }

    def _http(self, method: str, path: str, headers: dict, body: bytes | None):
        req = urllib.request.Request(self.cfg.key_server_url + path, data=body, method=method,
                                     headers=headers)
        with urllib.request.urlopen(req, timeout=self.cfg.timeout) as resp:  # noqa: S310 (HTTPS in prod)
            return resp.getcode(), resp.read()

    def fetch_key(self, module: str, ck: int, frag: str) -> dict:
        body = json.dumps({
            "module_id": module, "ck_version": ck, "frag_id": frag,
            "license_token": self.cfg.license_token,
        }).encode("utf-8")
        code, raw = self._http("POST", "/key", self._signed_headers("POST", "/key", body), body)
        data = json.loads(raw.decode("utf-8"))
        if not isinstance(data, dict):
            raise RuntimeError(f"risposta /key non valida (HTTP {code})")
        data["_http"] = code
        return data

    def post_tamper(self, tipo: str, dettaglio: str | None) -> None:
        body = json.dumps({"tipo": tipo, "dettaglio": dettaglio}).encode("utf-8")
        try:
            self._http("POST", "/tamper", self._signed_headers("POST", "/tamper", body), body)
        except Exception:  # noqa: BLE001 - reporting non bloccante
            pass

    def fetch_status(self) -> dict | None:
        body = json.dumps({"client_version": RUNTIME_VERSION, **_install_context()}).encode("utf-8")
        try:
            _code, raw = self._http("POST", "/status", self._signed_headers("POST", "/status", body), body)
        except Exception:  # noqa: BLE001 - disponibilità: il chiamante tiene l'ultimo stato noto
            return None
        try:
            data = json.loads(raw.decode("utf-8"))
        except Exception:  # noqa: BLE001
            return None
        if not isinstance(data, dict) or "signed_status" not in data:
            return None
        token = str(data["signed_status"])
        payload = crypto.verify_token(self.cfg.ed25519_public_key_pem, token)
        if payload is None:
            return None
        if payload.get("type") != "vaultcode/status/v1":
            return None
        if payload.get("install_uuid") != self.cfg.install_uuid:
            return None
        payload["_token"] = token
        return payload

    def fetch_revocations(self) -> list[str] | None:
        try:
            _code, raw = self._http("GET", "/revocations", {"Accept": "application/json"}, None)
        except Exception:  # noqa: BLE001
            return None
        try:
            data = json.loads(raw.decode("utf-8"))
        except Exception:  # noqa: BLE001
            return None
        if not isinstance(data, dict) or "signed_document" not in data:
            return None
        payload = crypto.verify_token(self.cfg.ed25519_public_key_pem, str(data["signed_document"]))
        if payload is None or "revoked" not in payload:
            return None
        return [str(r["license_id"]) for r in payload.get("revoked", []) if isinstance(r, dict) and "license_id" in r]
